### PR TITLE
Generate "Map" PathStructs and gNMI helpers

### PR DESCRIPTION
**Pull request by wenovus**
These are PathStructs that are unique to keyed-list types that refer to
    the entire map object representing the list.

    * This only pertains to compressed structs (will not affect uncompressed
      PathStruct generation).
    * Both non-wildcard and wildcard versions are generated. For Wildcard []map will
      be the return type.
    * Refactored the code to handle atomic lists as well without impacting ordered map
      generation. Child constructors for atomic lists are also cut off similar to ordered maps.

Tests are in following PR.

Comments to help with reviewing this change:
* Non-atomic lists have both regular PathStructs and "Map" PathStructs generated.
* Atomic (incl. ordered) lists have only "Map" PathStructs generated.
* Due to the above, only the corresponding `NodeData` are generated.

### CODE REVIEW

File: pathgen/gnmigen.go [hunk 1]
```diff
@@ -103,31 +109,35 @@ func GNMIGenerator(pathStructName string, dir *ygen.ParsedDirectory, node *NodeD
...
+				return err
+			}
+		}
+		return tmpl.Execute(&b, &tmplStruct)
 	}
 
-	if !generateConfigFunc(dir, node) {
-		return b.String(), nil
-	}
+	generate := func(tmplStruct gnmiStruct) error {
```
Comment (DanG100) [inline pathgen/gnmigen.go:123]: why is this a func, it's only used once

File: pathgen/pathgen.go [hunk 2]
```diff
@@ -767,45 +775,76 @@ func getNodeDataMap(ir *ygen.IR, fakeRootName, schemaStructPkgAccessor, pathStru
...
-				YANGPath:              field.YANGDetails.Path,
-				GoPathPackageName:     goPackageName(field.YANGDetails.RootElementModule, splitByModule, false, packageName, trimPrefix, packageSuffix),
-				DirectoryName:         field.YANGDetails.Path,
+				YANGTypeName: func() string {
+					if field.Flags != nil {
+						return field.Flags[yangTypeNameFlagKey]
+					} else {
+						return ""
+					}
+				}(),
```
Comment (DanG100) [inline pathgen/pathgen.go:800]: nit: don't inline this is hurts clarity

File: pathgen/pathgen.go [hunk 3]
```diff
@@ -767,45 +775,76 @@ func getNodeDataMap(ir *ygen.IR, fakeRootName, schemaStructPkgAccessor, pathStru
...
+				YANGPath:          field.YANGDetails.Path,
+				GoPathPackageName: goPackageName(field.YANGDetails.RootElementModule, splitByModule, false, packageName, trimPrefix, packageSuffix),
+				DirectoryName:     field.YANGDetails.Path,
 			}
 			switch {
-			case isLeaf:
-				nodeDataMap[pathStructName].YANGFieldName = fieldName
-				nodeDataMap[pathStructName].DirectoryName = dir.Path
-			case !isLeaf && field.YANGDetails.OrderedByUser: // "ordered-by user" lists
+			case !isLeaf && field.Type == ygen.ListNode && len(fieldDir.ListKeys) > 0 && !(fieldDir.CompressedTelemetryAtomic || field.YANGDetails.OrderedByUser):
```
Comment (DanG100) [inline pathgen/pathgen.go:806]: comment these cases or simplify the conditions

File: pathgen/gnmigen.go (hunk 1 above)
Comment (wenovus) [reply pathgen/gnmigen.go]: Removed.

File: pathgen/pathgen.go (hunk 2 above)
Comment (wenovus) [reply pathgen/pathgen.go]: done

File: pathgen/pathgen.go (hunk 3 above)
Comment (wenovus) [reply pathgen/pathgen.go]: Added a helper to improve readability

**Review by DanG100 — APPROVED** [review]

**Pull request merged**